src/Zigzag_persistence/include/gudhi/Zigzag_persistence.h

```diff
@@ -0,0 +1,751 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s):       Clément Maria
+ *
+ *    Copyright (C) 2021 Inria
+ *
+ *    Modification(s):
+ *      - 2023/05 Hannah Schreiber: Rework of the interface, reorganization and debug
+ *      - 2023/05 Hannah Schreiber: Addition of infinit bars
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+/**
+ * @file Zigzag_persistence.h
+ * @author Clément Maria, Hannah Schreiber
+ * @brief Contains the implementation of the @ref Gudhi::zigzag_persistence::Zigzag_persistence class.
+ */
+
+#ifndef ZIGZAG_PERSISTENCE_H_
+#define ZIGZAG_PERSISTENCE_H_
+
+#include <boost/tuple/tuple.hpp>
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/set.hpp>
+#include <boost/pending/disjoint_sets.hpp>
+#include <boost/pool/object_pool.hpp>
+#include <boost/timer/progress_display.hpp>
+
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <map>
+#include <list>
+#include <ostream>
+#include <set>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+#include <functional>
+
+#include <gudhi/Debug_utils.h>
+#include <gudhi/matrix.h>
+
+namespace Gudhi {
+namespace zigzag_persistence {
+
+/** \class Zigzag_persistence Zigzag_persistence.h gudhi/Zigzag_persistence.h
+ * \brief Class computating the zigzag persistent homology of a zigzag
+ * filtration. Algorithm based on \cite zigzag_reflection.
+ *
+ * \ingroup zigzag_persistence
+ *
+ * \details The type ZigzagComplex::Simplex_key counts the number of
+ * insertions and deletions of simplices, which may be large in zigzag persistence and require
+ * more than 32 bits of storage. The type used (int, long, etc) should be chosen in
+ * consequence. Simplex_key must be signed.
+ *
+ * \tparam ZigzagComplex Complex storing the current simplices.
+ * \tparam ZigzagPersistenceOptions Options for the matrix used to compute the persistence.
+ */
+template <typename ZigzagComplex,
+          typename ZigzagPersistenceOptions = Gudhi::persistence_matrix::Zigzag_options<> >
+class Zigzag_persistence 
+{
+ public:
+  using Complex = ZigzagComplex;                                /**< Complex type. */
+  using Options = ZigzagPersistenceOptions;                     /**< Matrix options */
+  /*** Types defined in the complex ***/
+  using Simplex_key = typename Complex::Simplex_key;            /**< Key type, must be signed. */
+  using Simplex_handle = typename Complex::Simplex_handle;      /**< Simplex ID type in the complex. */
+  using Vertex_handle = typename Complex::Vertex_handle;        /**< Vertex ID type in the complex. */
+  using Filtration_value = typename Complex::Filtration_value;  /**< Filtration value type. */
```

Comment:
How are filtration values interpreted exactly? Operations have an index (operation 0 is inserting this vertex, operation 1 is inserting this other vertex, etc). Is the filtration value just a map from indexes to reals (or whatever type) that we bundle with this structure for convenience, without any meaning? Or does it have an actual semantic (for instance say 2 simplices removed at the same filtration value may be removed in any order)?